external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaUtils.scala

```diff
@@ -144,4 +150,249 @@ object KafkaUtils {
     createStream[K, V, U, T](
       jssc.ssc, kafkaParams.toMap, Map(topics.mapValues(_.intValue()).toSeq: _*), storageLevel)
   }
+
+  /** A batch-oriented interface for consuming from Kafka.
+   * Starting and ending offsets are specified in advance,
+   * so that you can control exactly-once semantics.
+   * @param sc SparkContext object
+   * @param kafkaParams Kafka <a href="http://kafka.apache.org/documentation.html#configuration">
+   * configuration parameters</a>.
+   *   Requires "metadata.broker.list" or "bootstrap.servers" to be set with Kafka broker(s),
+   *   NOT zookeeper servers, specified in host1:port1,host2:port2 form.
+   * @param batch Each OffsetRange in the batch corresponds to a
+   *   range of offsets for a given Kafka topic/partition
+   */
+  @Experimental
+  def createRDD[
+    K: ClassTag,
+    V: ClassTag,
+    U <: Decoder[_]: ClassTag,
+    T <: Decoder[_]: ClassTag,
+    R: ClassTag] (
```

Comment:
The comment about unused R was referring to a prior version of the PR that
had a copy-pasted type-level R parameter even for the version that returned
RDD[(K, V)].  Github probably just got confused because the comment wasn't
attached to the particular line in question.

Pretty sure things are correct at this point.

On Tue, Feb 3, 2015 at 1:05 PM, Patrick Wendell <notifications@github.com>
wrote:

> In
> external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaUtils.scala
> <https://github.com/apache/spark/pull/3798#discussion_r24029463>:
>
> > +   * so that you can control exactly-once semantics.
> > +   * @param sc SparkContext object
> > +   * @param kafkaParams Kafka <a href="http://kafka.apache.org/documentation.html#configuration">
> > +   * configuration parameters</a>.
> > +   *   Requires "metadata.broker.list" or "bootstrap.servers" to be set with Kafka broker(s),
> > +   *   NOT zookeeper servers, specified in host1:port1,host2:port2 form.
> > +   * @param batch Each OffsetRange in the batch corresponds to a
> > +   *   range of offsets for a given Kafka topic/partition
> > +   */
> > +  @Experimental
> > +  def createRDD[
> > +    K: ClassTag,
> > +    V: ClassTag,
> > +    U <: Decoder[_]: ClassTag,
> > +    T <: Decoder[_]: ClassTag,
> > +    R: ClassTag] (
>
> Yeah, makes sense. But the comment here suggests R is not used, however,
> I see R being used in the return type. So that was my point.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/3798/files#r24029463>.
>